File: sdk/src/transact-with-poi.ts
/**
 * Phase 3d-full prototype — JoinSplit-with-PoI instruction builder.
 *
 * Today this only supports the (1, 2) variant. Other (N, M) variants need
 * their own compiled circuit + VK constants + a parallel verifier function
 * in the Rust contract; see `circuits/circom/joinsplit_with_poi.circom` for
 * the template and `TODOS.md` for the scale-out plan.
 *
 * The on-chain `transact_with_poi` (disc 26) is a *co-attestation* — it
 * verifies the augmented proof and emits an event tagging
 * `(nullifier, commitment_out_0, commitment_out_1, association_root)`.
 * Pair it with a regular `transact` (disc 13) in the same Solana
 * transaction; downstream consumers match the events on those fields to
 * conclude "this transact's input was clean."
 */

const INSTRUCTION_TRANSACT_WITH_POI = 26;
const GROTH16_PROOF_BYTES = 256;

/**
 * Build instruction data for `transact_with_poi` (disc 26) — 1x2 variant.
 *
 * Layout (418 bytes):
 *   disc(1) + n_inputs(1=1) + n_outputs(1=2) + proof(256)
 *   + merkle_root(32) + bound_params_hash(32) + nullifier(32)
 *   + commitment_out_0(32) + commitment_out_1(32)
 *
 * Note: unlike `transact` (which supports inline + buffer modes), the
 * prototype only supports inline proofs — the augmented Groth16 proof is
 * still 256 bytes, so the buffer mode optimization isn't needed yet.
 */
export function buildTransactWithPoIInstructionData(options: {
  /** 256-byte Groth16-with-PoI proof. */
  proofBytes: Uint8Array;
  merkleRoot: Uint8Array;
  boundParamsHash: Uint8Array;
  nullifier: Uint8Array;
  commitmentOut0: Uint8Array;
  commitmentOut1: Uint8Array;
}): Uint8Array {
  const expect = (buf: Uint8Array, len: number, name: string) => {
    if (buf.length !== len) {
      throw new Error(`${name} must be ${len} bytes; got ${buf.length}`);
    }
  };
  expect(options.proofBytes, GROTH16_PROOF_BYTES, "proofBytes");
  expect(options.merkleRoot, 32, "merkleRoot");
  expect(options.boundParamsHash, 32, "boundParamsHash");
  expect(options.nullifier, 32, "nullifier");
  expect(options.commitmentOut0, 32, "commitmentOut0");
  expect(options.commitmentOut1, 32, "commitmentOut1");

  // 1 + 1 + 1 + 256 + 32 + 32 + 32 + 32 + 32 = 419
  const out = new Uint8Array(1 + 1 + 1 + GROTH16_PROOF_BYTES + 32 * 5);
  let off = 0;
  out[off++] = INSTRUCTION_TRANSACT_WITH_POI;
  out[off++] = 1; // n_inputs
  out[off++] = 2; // n_outputs
  out.set(options.proofBytes, off);
  off += GROTH16_PROOF_BYTES;
  out.set(options.merkleRoot, off);
  off += 32;
  out.set(options.boundParamsHash, off);
  off += 32;
  out.set(options.nullifier, off);
  off += 32;
  out.set(options.commitmentOut0, off);
  off += 32;
  out.set(options.commitmentOut1, off);
  return out;
}

File: contracts/programs/utxopia/src/instructions/transact_with_poi.rs
//! Transact-with-PoI co-attestation (Phase 3d-full prototype).
//!
//! ## Framing
//!
//! Rather than cloning the full ~400-line `transact` handler, this
//! instruction is a co-attestation: it verifies a JoinSplit-with-PoI Groth16
//! proof for the (1, 2) variant and emits a tagging event. The actual
//! state-change work (nullifier PDAs, leaf insertion, stealth announcements,
//! sender memos) still flows through the regular `transact` (disc 13) ix in
//! the same Solana transaction.
//!
//! Downstream consumers (CEXes, auditors) pair the two events on matching
//! `(merkle_root, bound_params_hash, nullifier, commitments_out)` tuples to
//! conclude "this transact was PoI-clean as of `association_root`."
//!
//! The verifier enforces:
//!   1. The Groth16 proof binds the public inputs.
//!   2. `merkle_root` is an active root in the commitment tree (so attackers
//!      can't fabricate against an old or forked root).
//!   3. `association_root` matches the current `AssociationSet` PDA root.
//!
//! Combined, this means: any pair of (transact + transact_with_poi) events
//! with matching nullifier + commitments is a public, on-chain assertion
//! that the spent note's commitment was in the curated clean set.
//!
//! ## Instruction Data Layout (only 1x2 variant — Phase 3d-full prototype)
//!   - [0]      n_inputs         (u8, MUST be 1)
//!   - [1]      n_outputs        (u8, MUST be 2)
//!   - [2..258] proof            ([u8; 256] inline only — no buffer mode)
//!   - [258..290]  merkle_root        ([u8; 32])
//!   - [290..322]  bound_params_hash  ([u8; 32])
//!   - [322..354]  nullifier          ([u8; 32])
//!   - [354..386]  commitment_out_0   ([u8; 32])
//!   - [386..418]  commitment_out_1   ([u8; 32])
//! Total: 418 bytes.
//!
//! ## Accounts
//!   0. association_set    (read, PDA — pinned with current_root)
//!   1. commitment_tree    (read, validates merkle_root is active)
//!   2. payer              (signer, pays the rent if any)
//!
//! ## Scaling out
//! Other (N, M) variants need: their own compiled circuit + zkey, their own
//! VK constants in `utils/groth16.rs`, and a parameterized version of this
//! handler that picks the right `verify_groth16_transact_NxM_with_poi_proof`
//! function. See TODOS.md → "Phase 3d-full scale-out".

use pinocchio::{account_info::AccountInfo, program_error::ProgramError, pubkey::Pubkey, ProgramResult};

use crate::error::UTXOpiaError;
use crate::state::{AssociationSet, CommitmentTree};
use crate::utils::groth16::GROTH16_PROOF_SIZE;
use crate::utils::validate_program_owner;

/// Total instruction data size for the 1x2 prototype.
/// header(2) + proof(256) + roots(32+32) + nullifier(32) + commitments(32*2) = 418
const IX_DATA_SIZE: usize = 2 + GROTH16_PROOF_SIZE + 32 + 32 + 32 + 32 + 32;

pub fn process_transact_with_poi(
    program_id: &Pubkey,
    accounts: &[AccountInfo],
    data: &[u8],
) -> ProgramResult {
    if data.len() < IX_DATA_SIZE {
        return Err(ProgramError::InvalidInstructionData);
    }
    if accounts.len() < 3 {
        return Err(ProgramError::NotEnoughAccountKeys);
    }

    // ── Parse instruction data ────────────────────────────────────────
    let n_inputs = data[0];
    let n_outputs = data[1];

    // 1x2 prototype: hard-fail any other variant. Future scale-out will
    // dispatch on (n_inputs, n_outputs) to the right VK + verifier.
    if n_inputs != 1 || n_outputs != 2 {
        return Err(ProgramError::InvalidInstructionData);
    }

    let mut off = 2;
    let proof_bytes = &data[off..off + GROTH16_PROOF_SIZE];
    off += GROTH16_PROOF_SIZE;
    let merkle_root: &[u8; 32] = data[off..off + 32].try_into().unwrap();
    off += 32;
    let bound_params_hash: &[u8; 32] = data[off..off + 32].try_into().unwrap();
    off += 32;
    let nullifier: &[u8; 32] = data[off..off + 32].try_into().unwrap();
    off += 32;
    let commitment_out_0: &[u8; 32] = data[off..off + 32].try_into().unwrap();
    off += 32;
    let commitment_out_1: &[u8; 32] = data[off..off + 32].try_into().unwrap();

    // ── Validate accounts ─────────────────────────────────────────────
    let association_info = &accounts[0];
    let commitment_tree_info = &accounts[1];
    let payer = &accounts[2];

    validate_program_owner(association_info, program_id)?;
    validate_program_owner(commitment_tree_info, program_id)?;
    if !payer.is_signer() {
        return Err(ProgramError::MissingRequiredSignature);
    }

    // ── Pin the association root + check it's not paused ─────────────
    let association_data = association_info.try_borrow_data()?;
    let set = AssociationSet::from_bytes(&association_data)?;
    if set.status != 0 {
        return Err(UTXOpiaError::PoolPaused.into());
    }
    let association_root = set.current_root;

    // ── Validate the JoinSplit merkle_root is a known/active tree root.
    //    This is the same check transact does — prevents an attacker from
    //    forging a proof against an old or rotated-out root.
    {
        let tree_data = commitment_tree_info.try_borrow_data()?;
        let tree = CommitmentTree::from_bytes(&tree_data)?;
        if !tree.is_valid_root(merkle_root) {
            return Err(UTXOpiaError::InvalidMerkleProof.into());
        }
    }

    // ── Verify the Groth16-with-PoI proof ─────────────────────────────
    crate::utils::groth16::verify_groth16_transact_1x2_with_poi_proof(
        proof_bytes,
        merkle_root,
        bound_params_hash,
        nullifier,
        commitment_out_0,
        commitment_out_1,
        &association_root,
    )?;

    // ── Emit the co-attestation event ─────────────────────────────────
    // Downstream consumers pair this with a transact event on the matching
    // (nullifier, commitments_out) to conclude "that transact was clean."
    crate::utils::events::emit_transact_with_poi(
        &association_root,
        nullifier,
        commitment_out_0,
        commitment_out_1,
        set.version,
    );

    Ok(())
}
